File: kernel/nodejs/src/directory-lock.ts
import { NOOP_LOGGER, type Logger } from "@telorun/sdk";
import * as fs from "fs/promises";
import * as os from "os";
import * as path from "path";

/**
 * The cross-process lock guarding mutation of a shared cache directory.
 *
 * Two consumers hold it for the same reason — several Telo processes (and, in
 * one process, several concurrent resolutions) may decide to populate the same
 * directory at the same moment: `NpmControllerLoader` around its install root,
 * and `ModuleArtifact` around a module's extracted layers. One implementation
 * rather than two, because the reclaim rule below is subtle enough that two
 * copies would drift.
 */

/**
 * A held lock is refreshed (its mtime bumped) every {@link LOCK_HEARTBEAT_MS}
 * by the holder. Staleness is judged purely by mtime age — a lock older than
 * this means the holder stopped heartbeating (crashed, was killed, or its
 * container vanished), so it is safe to reclaim. This deliberately does NOT
 * probe the recorded PID for liveness: PID identity is meaningless across
 * container restarts and PID namespaces (deterministic PID reuse makes an
 * unrelated process look like the dead holder on the same hostname), which is
 * exactly what deadlocked container boots. The `{pid, host}` in the lock body
 * is diagnostics for humans, never a reclaim signal. Must be comfortably
 * larger than the heartbeat interval so a briefly-descheduled holder (GC
 * pause, busy event loop) is not reclaimed out from under itself.
 */
const LOCK_STALE_MS = 30_000;

/** How often the holder refreshes the lock mtime while `fn` runs. Well under
 *  {@link LOCK_STALE_MS} so several heartbeats are missed before a live holder
 *  is ever judged stale. */
const LOCK_HEARTBEAT_MS = 5_000;

/**
 * Total wall-clock cap for waiting on the lock — enough for a slow first
 * populate on a peer process to finish, short enough that a genuinely
 * deadlocked CI job fails loudly rather than hanging for hours. The retry
 * interval trades wakeup latency vs. wasted polls; 500ms is well below the
 * lock holder's typical hold time.
 */
const LOCK_WAIT_MAX_MS = 5 * 60_000;
const LOCK_RETRY_MS = 500;

/** After this long waiting on a lock, emit one line so a slow wait is visible
 *  instead of looking like a silent hang. */
const LOCK_WAIT_NOTICE_MS = 2_000;

const sleep = (ms: number): Promise<void> => new Promise((resolve) => setTimeout(resolve, ms));

/**
 * In-process serialization ahead of the filesystem lock.
 *
 * Without it, N concurrent callers for the same directory all reach the fs lock,
 * N-1 sit in the retry loop, and each crossing {@link LOCK_WAIT_NOTICE_MS}
 * prints the wait notice — a notice whose whole point is "another *process* is
 * populating this", printed when the holder is us. (`telo install` fanning 52
 * controllers out through one `Promise.allSettled` printed 51 of them.)
 *
 * Queuing here means exactly one caller per process reaches the fs lock, so the
 * notice regains its cross-process meaning and the losers do no I/O at all. The
 * fs lock is untouched and still provides the cross-process guarantee.
 */
const localQueues = new Map<string, Promise<unknown>>();

function withLocalQueue<T>(dir: string, fn: () => Promise<T>): Promise<T> {
  const prev = localQueues.get(dir) ?? Promise.resolve();
  // Run on both settle paths: one failure must not wedge the queue.
  const run = prev.then(fn, fn);
  // The stored tail never rejects — a failure neither poisons followers nor
  // surfaces as an unhandled rejection on the chain copy.
  const tail = run.then(
    () => {},
    () => {},
  );
  localQueues.set(dir, tail);
  // Drop the entry once nothing further is queued, so a long-lived process that
  // touches many directories doesn't retain a promise per directory forever.
  void tail.then(() => {
    if (localQueues.get(dir) === tail) localQueues.delete(dir);
  });
  return run;
}

/**
 * Acquire the lock for `dir` and run `fn` under it: first the in-process queue
 * above, then the cross-process filesystem lock. `label` names the operation in
 * the wait notice and timeout error (e.g. "controller install", "module layer").
 */
export async function withDirectoryLock<T>(
  dir: string,
  label: string,
  fn: () => Promise<T>,
  log: Logger = NOOP_LOGGER,
): Promise<T> {
  return withLocalQueue(dir, () => withFileLock(dir, label, fn, log));
}

/**
 * Acquire a process-portable lock on `<dir>/.lock` and execute fn while
 * holding it. `fs.open(path, 'wx')` is atomic on POSIX and Windows, so
 * concurrent processes serialize naturally.
 *
 * Liveness is a heartbeat: the holder bumps the lock's mtime every
 * {@link LOCK_HEARTBEAT_MS} while `fn` runs, and a waiter reclaims a lock whose
 * mtime is older than {@link LOCK_STALE_MS} (holder crashed/killed/vanished).
 * mtime age is the *only* reclaim signal — the recorded `{pid, host}` is
 * diagnostics, never probed for liveness, because PID identity is unreliable
 * across container restarts and PID namespaces (the failure that deadlocked
 * container boots). Reclaim is via atomic rename to a unique tombstone so two
 * waiters that both see the lock stale can't both win.
 *
 * The lock guards *writes* into the directory. It does NOT serialize reads of
 * already-populated content — those run lock-free against a stable tree.
 */
async function withFileLock<T>(
  dir: string,
  label: string,
  fn: () => Promise<T>,
  log: Logger,
): Promise<T> {
  const lockPath = path.join(dir, ".lock");

  await fs.mkdir(dir, { recursive: true });

  const lockBody = JSON.stringify({ pid: process.pid, host: os.hostname(), startedAt: Date.now() });
  let handle: import("fs/promises").FileHandle | null = null;
  const waitedSince = Date.now();
  let noticed = false;
  while (true) {
    try {
      handle = await fs.open(lockPath, "wx");
      await handle.writeFile(lockBody);
      break;
    } catch (err: any) {
      if (err?.code !== "EEXIST") throw err;
      // Lock exists. Reclaim it only if its heartbeat has gone silent.
      if (await reclaimIfStale(lockPath)) continue;
      const waited = Date.now() - waitedSince;
      if (waited > LOCK_WAIT_MAX_MS) {
        throw new Error(
          `[telo] timed out waiting for ${label} lock at ${lockPath} ` +
            `(held >${LOCK_WAIT_MAX_MS / 60_000} min with a live heartbeat). ` +
            `Inspect the lock file or remove it manually if no other Telo process is running.`,
        );
      }
      if (!noticed && waited > LOCK_WAIT_NOTICE_MS) {
        noticed = true;
        log.info(`waiting for ${label} lock`, { "telo.lock.path": lockPath });
      }
      await sleep(LOCK_RETRY_MS);
    }
  }

  // Keep the lock fresh while `fn` runs so a slow-but-live operation is never
  // reclaimed. `unref` so the heartbeat can't by itself keep the process alive.
  const heartbeat = setInterval(() => {
    const now = new Date();
    fs.utimes(lockPath, now, now).catch(() => {});
  }, LOCK_HEARTBEAT_MS);
  heartbeat.unref?.();

  try {
    return await fn();
  } finally {
    clearInterval(heartbeat);
    // The fd close races nothing important: if it fails, the FD is reaped on
    // process exit. The unlink is the dangerous one — a non-ENOENT failure
    // (permissions, read-only mount) means every subsequent kernel waits
    // LOCK_STALE_MS before reclaiming. Surface it so the cause is visible
    // rather than hiding behind a silent hang.
    await handle!.close().catch(() => {});
    try {
      await fs.rm(lockPath, { force: true });
    } catch (err: any) {
      if (err?.code !== "ENOENT") {
        log.warn(`failed to release ${label} lock`, { "telo.lock.path": lockPath }, { error: err });
      }
    }
  }
}

/**
 * If the lock at `lockPath` is stale (mtime older than {@link LOCK_STALE_MS}, so
 * its holder stopped heartbeating), atomically claim and remove it and return
 * true; otherwise return false. The claim is a `rename` to a unique tombstone:
 * `rename` is atomic and fails for all but one racer, so two processes that
 * both observe the same stale lock cannot both reclaim it — the loser's rename
 * throws ENOENT (the file is already gone) and it simply retries the open.
 */
async function reclaimIfStale(lockPath: string): Promise<boolean> {
  let stat: import("fs").Stats;
  try {
    stat = await fs.stat(lockPath);
  } catch (err: any) {
    // Race: lock vanished while we inspected it. Retry the open immediately.
    if (err?.code === "ENOENT") return true;
    throw err;
  }
  if (Date.now() - stat.mtimeMs < LOCK_STALE_MS) return false;

  // Stale — the holder's heartbeat is silent. Claim via atomic rename; only one
  // racer wins, the rest get ENOENT and fall back to retrying the open.
  const tombstone = `${lockPath}.stale.${process.pid}.${stat.mtimeMs}`;
  try {
    await fs.rename(lockPath, tombstone);
  } catch (err: any) {
    if (err?.code === "ENOENT") return true; // another waiter reclaimed it first
    throw err;
  }
  await fs.rm(tombstone, { force: true });
  return true;
}

/**
 * Internals exposed for the lock's own tests (acquire/release, stale reclaim,
 * heartbeat, in-process queuing). Not part of the kernel's public API.
 */
export const __testing__ = {
  reclaimIfStale,
  LOCK_STALE_MS,
  LOCK_HEARTBEAT_MS,
};

File: kernel/rust/src/directory_lock.rs
//! The cross-process lock guarding mutation of a shared cache directory.
//! Mirrors `../../nodejs/src/directory-lock.ts`.
//!
//! **Protocol-compatible with the Node lock, not merely similar**: both kernels
//! populate one `.telo` cache, so a lock only one of them can see is no lock.
//! Same file (`<dir>/.lock`), same exclusive create, same JSON body, same
//! mtime heartbeat and staleness threshold, same rename-to-tombstone reclaim —
//! an OS `flock` would be invisible to the Node kernel and the two would race.
//!
//! Node's in-process queue ahead of the file lock has no counterpart: it
//! serializes concurrent async callers inside one process, and this kernel is
//! synchronous, so there is never a second caller to queue.

use std::fs::{self, File, OpenOptions};
use std::io::{ErrorKind, Write};
use std::path::{Path, PathBuf};
use std::sync::mpsc::{self, RecvTimeoutError};
use std::thread;
use std::time::{Duration, Instant, SystemTime, UNIX_EPOCH};

use crate::error::KernelError;

/// A lock whose mtime is older than this has lost its holder's heartbeat and is
/// reclaimed. mtime age is the only reclaim signal: the recorded pid is
/// meaningless across containers and PID namespaces.
const LOCK_STALE: Duration = Duration::from_secs(30);
/// How often a holder bumps the lock's mtime.
const LOCK_HEARTBEAT: Duration = Duration::from_secs(5);
/// Total wait before giving up on a lock with a live heartbeat.
const LOCK_WAIT_MAX: Duration = Duration::from_secs(5 * 60);
const LOCK_RETRY: Duration = Duration::from_millis(500);
/// After this long waiting, say so once, so a slow wait is not a silent hang.
const LOCK_WAIT_NOTICE: Duration = Duration::from_secs(2);

#[derive(Debug, thiserror::Error)]
pub enum LockError {
    #[error("[telo] {operation} for {label} lock at {path}: {source}")]
    Io {
        operation: &'static str,
        label: String,
        path: PathBuf,
        source: std::io::Error,
    },
    #[error(
        "[telo] timed out waiting for {label} lock at {path} (held >{minutes} min with a live heartbeat). \
         Inspect the lock file or remove it manually if no other Telo process is running."
    )]
    TimedOut {
        label: String,
        path: PathBuf,
        minutes: u64,
    },
}

impl From<LockError> for KernelError {
    fn from(err: LockError) -> Self {
        KernelError::new("ERR_DIRECTORY_LOCK_FAILED", err.to_string())
    }
}

/// Acquire the lock for `dir` and run `body` under it. `label` names the
/// operation in the wait notice and the timeout error.
///
/// The lock guards writes into the directory; reads of already-populated
/// content run lock-free.
pub fn with_directory_lock<T, E: From<LockError>>(
    dir: &Path,
    label: &str,
    body: impl FnOnce() -> Result<T, E>,
) -> Result<T, E> {
    let lock_path = dir.join(".lock");
    let io = |operation, source| LockError::Io {
        operation,
        label: label.to_string(),
        path: lock_path.clone(),
        source,
    };
    fs::create_dir_all(dir).map_err(|err| io("creating the directory", err))?;

    // Monotonic: a wall clock stepped backwards would otherwise stretch the wait
    // past its bound, and one stepped forwards would time out a live holder.
    let waited_since = Instant::now();
    let mut noticed = false;
    let mut held = loop {
        match OpenOptions::new().write(true).create_new(true).open(&lock_path) {
            Ok(mut file) => {
                // Owned from the moment the file exists, so a failed write still
                // removes it rather than leaving every later kernel to wait it out.
                let held = HeldLock {
                    path: lock_path.clone(),
                    label: label.to_string(),
                    heartbeat: None,
                };
                file.write_all(lock_body().as_bytes())
                    .map_err(|err| io("writing", err))?;
                break held;
            }
            Err(err) if err.kind() == ErrorKind::AlreadyExists => {
                if reclaim_if_stale(&lock_path).map_err(|err| io("reclaiming a stale lock", err))? {
                    continue;
                }
                let waited = waited_since.elapsed();
                if waited > LOCK_WAIT_MAX {
                    return Err(LockError::TimedOut {
                        label: label.to_string(),
                        path: lock_path,
                        minutes: LOCK_WAIT_MAX.as_secs() / 60,
                    }
                    .into());
                }
                if !noticed && waited > LOCK_WAIT_NOTICE {
                    noticed = true;
                    eprintln!("telo: waiting for {label} lock at {}", lock_path.display());
                }
                thread::sleep(LOCK_RETRY);
            }
            Err(err) => return Err(io("creating", err).into()),
        }
    };

    held.heartbeat = Some(Heartbeat::start(lock_path.clone(), label.to_string()));
    // Released by `held`'s drop on every exit, a panicking body included.
    body()
}

/// A lock this process holds: dropping it stops the heartbeat and removes the
/// file, however the holder's work ended.
struct HeldLock {
    path: PathBuf,
    label: String,
    heartbeat: Option<Heartbeat>,
}

impl Drop for HeldLock {
    fn drop(&mut self) {
        if let Some(heartbeat) = self.heartbeat.take() {
            heartbeat.stop();
        }
        // A lock left behind makes every later kernel wait out LOCK_STALE, so a
        // failure to remove it is reported rather than dropped.
        if let Err(err) = fs::remove_file(&self.path) {
            if err.kind() != ErrorKind::NotFound {
                eprintln!(
                    "telo: failed to release {} lock at {}: {err}",
                    self.label,
                    self.path.display()
                );
            }
        }
    }
}

/// `{pid, host, startedAt}` — diagnostics for a human, never a reclaim signal.
fn lock_body() -> String {
    let started_at = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|elapsed| elapsed.as_millis())
        .unwrap_or(0);
    serde_json::json!({
        "pid": std::process::id(),
        "host": hostname(),
        "startedAt": started_at,
    })
    .to_string()
}

/// The host name for the lock body. The standard library cannot ask the OS, so
/// the usual environment and file are consulted, and an unnamed host records
/// that it is unnamed.
fn hostname() -> String {
    ["HOSTNAME", "COMPUTERNAME"]
        .iter()
        .find_map(|key| std::env::var(key).ok().filter(|v| !v.trim().is_empty()))
        .or_else(|| {
            fs::read_to_string("/etc/hostname")
                .ok()
                .map(|name| name.trim().to_string())
                .filter(|name| !name.is_empty())
        })
        .unwrap_or_else(|| "unknown".to_string())
}

/// If the lock is stale, claim it by an atomic rename to a unique tombstone —
/// which only one racer wins — and remove it. `true` when the caller should
/// retry the create immediately.
fn reclaim_if_stale(lock_path: &Path) -> std::io::Result<bool> {
    let modified = match fs::metadata(lock_path).and_then(|meta| meta.modified()) {
        Ok(modified) => modified,
        Err(err) if err.kind() == ErrorKind::NotFound => return Ok(true),
        Err(err) => return Err(err),
    };
    // A future mtime reads as fresh, as it does in Node.
    let age = SystemTime::now().duration_since(modified).unwrap_or_default();
    if age < LOCK_STALE {
        return Ok(false);
    }
    let modified_ms = modified
        .duration_since(UNIX_EPOCH)
        .map(|since| since.as_millis())
        .unwrap_or(0);
    let tombstone = PathBuf::from(format!(
        "{}.stale.{}.{modified_ms}",
        lock_path.display(),
        std::process::id()
    ));
    match fs::rename(lock_path, &tombstone) {
        Ok(()) => {}
        Err(err) if err.kind() == ErrorKind::NotFound => return Ok(true),
        Err(err) => return Err(err),
    }
    match fs::remove_file(&tombstone) {
        Err(err) if err.kind() != ErrorKind::NotFound => Err(err),
        _ => Ok(true),
    }
}

/// Keeps the lock's mtime fresh while the body runs, so a slow-but-live holder
/// is never reclaimed.
struct Heartbeat {
    stop: mpsc::Sender<()>,
    thread: thread::JoinHandle<()>,
}

impl Heartbeat {
    fn start(lock_path: PathBuf, label: String) -> Self {
        let (stop, stopped) = mpsc::channel::<()>();
        let thread = thread::spawn(move || {
            let mut reported = false;
            while let Err(RecvTimeoutError::Timeout) = stopped.recv_timeout(LOCK_HEARTBEAT) {
                let bumped = File::options()
                    .write(true)
                    .open(&lock_path)
                    .and_then(|file| file.set_modified(SystemTime::now()));
                // Reported once: a peer that judges this lock stale will take
                // it, and every later beat would repeat the same line.
                if let Err(err) = bumped {
                    if !reported {
                        reported = true;
                        eprintln!(
                            "telo: failed to refresh {label} lock at {}: {err}",
                            lock_path.display()
                        );
                    }
                }
            }
        });
        Self { stop, thread }
    }

    fn stop(self) {
        // The receiver lives until the thread exits, so a send can only fail
        // once the thread is already gone.
        let _ = self.stop.send(());
        if self.thread.join().is_err() {
            eprintln!("telo: the lock heartbeat thread panicked");
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn releases_the_lock_when_the_body_panics() {
        let dir = tempfile::tempdir().unwrap();
        let lock = dir.path().join(".lock");
        let path = dir.path().to_path_buf();
        let outcome = std::panic::catch_unwind(move || {
            with_directory_lock::<(), LockError>(&path, "module layer", || panic!("the body failed"))
        });
        assert!(outcome.is_err());
        assert!(!lock.exists(), "a panicking holder left its lock behind");
    }

    /// A lock file exactly as the Node kernel writes it.
    fn node_lock(dir: &Path) -> PathBuf {
        let path = dir.join(".lock");
        fs::write(&path, r#"{"pid":4242,"host":"node-kernel","startedAt":1757771234567}"#).unwrap();
        path
    }

    #[test]
    fn a_lock_held_by_the_node_kernel_blocks_until_released() {
        let dir = tempfile::tempdir().unwrap();
        let lock = node_lock(dir.path());
        let release = thread::spawn(move || {
            thread::sleep(Duration::from_millis(1200));
            fs::remove_file(lock).unwrap();
        });
        let started = Instant::now();
        with_directory_lock::<_, LockError>(dir.path(), "module layer", || Ok(())).unwrap();
        assert!(
            started.elapsed() >= Duration::from_millis(1200),
            "entered after {:?}, while the Node holder still held the lock",
            started.elapsed()
        );
        release.join().unwrap();
    }

    #[test]
    fn a_stale_node_lock_is_reclaimed() {
        let dir = tempfile::tempdir().unwrap();
        let lock = node_lock(dir.path());
        File::options()
            .write(true)
            .open(&lock)
            .unwrap()
            .set_modified(SystemTime::now() - Duration::from_secs(60))
            .unwrap();
        let started = Instant::now();
        with_directory_lock::<_, LockError>(dir.path(), "module layer", || Ok(())).unwrap();
        assert!(started.elapsed() < LOCK_RETRY, "a stale lock must not be waited on");
    }

    /// The Node kernel reads the lock this kernel writes as its own: a JSON body
    /// at `<dir>/.lock` while held, and no file once released.
    #[test]
    fn writes_the_node_lock_format_and_releases_it() {
        let dir = tempfile::tempdir().unwrap();
        let lock = dir.path().join(".lock");
        let body = with_directory_lock::<_, LockError>(dir.path(), "module layer", || {
            Ok(fs::read_to_string(&lock).unwrap())
        })
        .unwrap();
        let body: serde_json::Value = serde_json::from_str(&body).unwrap();
        assert_eq!(body["pid"], std::process::id());
        assert!(body["host"].is_string() && body["startedAt"].is_u64(), "{body}");
        assert!(!lock.exists());
    }
}
